src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/CorrelationVector.cs

```diff
@@ -0,0 +1,207 @@
+                        size = baseVector.Length + 1 + (int)Math.Log10(next) + 1;
+                        if (size > CorrelationVector.MaxVectorLength)
+                        {
+                            this.PreviousBase = GetBaseFromVector(this.baseVector);
+                            this.baseVector = CorrelationVector.GetBase(isReset: true);
+                        }
+                    }
+                }
+            }
+            while (snapshot != Interlocked.CompareExchange(ref this.extension, next, snapshot));
+
+            return string.Concat(this.baseVector, ElementChar, next);
+        }
+
+        /// <summary>
+        /// Determines whether two instances of the <see cref="CorrelationVector"/> class
+        /// are equal. 
+        /// </summary>
+        /// <param name="correlationVector">
+        /// The correlation vector you want to compare with the current correlation vector.
+        /// </param>
+        /// <returns>
+        /// True if the specified correlation vector is equal to the current correlation
+        /// vector; otherwise, false.
+        /// </returns>
+        public bool Equals(CorrelationVector correlationVector)
+        {
+            return string.Equals(this.Value, correlationVector.Value, StringComparison.Ordinal);
+        }
+
+        /// <summary>
+        /// Creates a new correlation vector by extending an existing value. This should be
+        /// done at the entry point of an operation.
+        /// </summary>
+        /// <param name="correlationVector">
+        /// Taken from the message header.
+        /// </param>
+        /// <returns>A new correlation vector extended from the current vector.</returns>
+        public static CorrelationVector Extend(string correlationVector)
+        {
+            // 2 accounts for the ".0" at the end of the new CorrelationVector
+            int size = ((correlationVector == null) ? 0 : correlationVector.Length) + 2;
+
+            if (size > CorrelationVector.MaxVectorLength)
+            {
+                return new CorrelationVector()
+                {
+                    baseVector = CorrelationVector.GetBase(isReset: true),
+                    extension = 0,
+                    PreviousBase = GetBaseFromVector(correlationVector)
+                };
+            }
+
+            return new CorrelationVector()
+            {
+                baseVector = correlationVector,
+                extension = 0
+            };
+        }
+
+        /// <summary>
+        /// TBD - describe this thing
+        /// </summary>
+        /// <param name="correlationVector">
+        /// Taken from the message header.
+        /// </param>
+        /// <returns>A new correlation vector extended from the current vector.</returns>
+        public static CorrelationVector Spin(string correlationVector)
+        {
+            ulong spinElement = GetRandomUnsignedLong();
```

Comment:
Curious- why is the element that gets added by Spin 8-bytes while the element that gets updated by an Increment is only 4-bytes?